[FIX]: Fixing issue seen in service update with network add and remove

abhinandanpb opened the pull request:
The PR contains fix and test cases to address the issue : 
https://github.com/moby/moby/issues/32629

The root cause of the issue was with some parts of the code assuming to address only single network attachment for service. The PR addresses issue to handle combination of sequence like network add delete in the same service update or over multiple service updates.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "fix32629" git@github.com:abhinandanpb/swarmkit.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353675464
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

aboch commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -303,38 +324,84 @@ func OnInit(options *ServiceAllocationOpts) {
 
+	// Always prefer NetworkAttachmentConfig in the TaskSpec
+	specNetworks := s.Spec.Task.Networks
+	if len(specNetworks) == 0 && s != nil && len(s.Spec.Networks) != 0 {
```
if `s` was nil, previous line would have paniced.
Is it possible it is nil when this code is called ?

aboch commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -303,38 +324,84 @@ func OnInit(options *ServiceAllocationOpts) {
+}
 
+// IsServiceToBeAllocated returns true if the passed service needs to have network resources allocated/updated.
+func (na *NetworkAllocator) IsServiceToBeAllocated(s *api.Service, flags []func(*ServiceAllocationOpts)) bool {
```
please make `flags` parameter variadic as done in `IsServiceAllocated`

aboch commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -186,12 +187,28 @@ func (na *NetworkAllocator) ServiceAllocate(s *api.Service) (err error) {
-		if err = na.allocateVIP(eAttach); err != nil {
-			return
+	// Allocate VIPs for all the pre-populated endpoint attachments
+	eVIPs := s.Endpoint.VirtualIPs[:0]
```
What is the advantage of defining `eVIPs` this way compared to ` var eVIPs []*api.Endpoint_VirtualIP` ?
Is it so you do not need to care what types the elements are and yo udo not need to import their package ?

abhinandanpb replied:
@aboch not really. The advantage is mentioned here https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating
I was just following that convention

aboch commented:
looks good to me

aaronlehmann commented on manager/allocator/networkallocator/networkallocator_test.go:
```diff
@@ -718,6 +718,98 @@ func TestServiceUpdate(t *testing.T) {
+	err = na.ServiceAllocate(s)
+	assert.NoError(t, err)
+	assert.Equal(t, true, na.IsServiceAllocated(s))
+	assert.Equal(t, 1, len(s.Endpoint.VirtualIPs))
```
```go
assert.True(t, na.IsServiceAllocated(s))
assert.Len(t, s.Endpoint.VirtualIPs, 1)
```

and so on for the other assertions below

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -303,38 +324,84 @@ func OnInit(options *ServiceAllocationOpts) {
+}
 
+// IsServiceToBeAllocated returns true if the passed service needs to have network resources allocated/updated.
+func (na *NetworkAllocator) IsServiceToBeAllocated(s *api.Service, flags ...func(*ServiceAllocationOpts)) bool {
```
Maybe it could be named `ServiceNeedsAllocation`?

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -303,38 +324,84 @@ func OnInit(options *ServiceAllocationOpts) {
 	// If endpoint mode is VIP and allocator does not have the
 	// service in VIP allocated set then it is not allocated.
-	if (len(s.Spec.Task.Networks) != 0 || len(s.Spec.Networks) != 0) &&
+	if (len(specNetworks) != 0) &&
```
nit: The parens around `len(specNetworks) != 0` aren't needed

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -303,38 +324,84 @@ func OnInit(options *ServiceAllocationOpts) {
+			return true
+		}
+
+		var match bool
```
It looks like `match` is reset on every iteration of the loop and is only used inside the loop, so this declaration could be removed and the loop could start with `match := false`.

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -303,38 +324,84 @@ func OnInit(options *ServiceAllocationOpts) {
+	//If the spec no longer has networks attached and has a vip allocated
+	//from previous spec the service needs to updated
+	if s.Endpoint != nil {
+		var match bool
```
As above, this declaration could be removed and replaced with `match := false` inside the loop. Or we could use a label and `continue`:

```go
vipsLoop:
		for _, vip := range s.Endpoint.VirtualIPs {
			for _, net := range specNetworks {
				if vip.NetworkID == net.Target {
					continue vipsLoop
				}
			}
			return true
		}
	}
```

I find this a little easier to follow, but opinions differ :)

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -303,38 +324,84 @@ func OnInit(options *ServiceAllocationOpts) {
 func (na *NetworkAllocator) IsServiceAllocated(s *api.Service, flags ...func(*ServiceAllocationOpts)) bool {
-	var options ServiceAllocationOpts
+	return !na.IsServiceToBeAllocated(s, flags...)
+}
```
Rather than define both the function and its inverse, let's choose one or the other.

abhinandanpb replied:
The only reason to retain the api was to keep it in sync with all the other objects which has similar apis. I can change it if its not necessary.

aaronlehmann replied:
Feel free to change it. Personally, I don't really like the name `IsServiceAllocated`, since the function is really checking if the service *needs* allocation, not if it is allocated. I don't mind whether the function returns whether the service needs allocation or does not need allocation, as long as the name is clear.

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -303,38 +324,84 @@ func OnInit(options *ServiceAllocationOpts) {
+			return true
+		}
+
+		if s.Endpoint.VirtualIPs == nil {
```
`if len(s.Endpoint.VirtualIPs) == 0`

It's possible for the slice to be allocated, but have no items. It's better to avoid assumptions about whether the slice would be allocated in this case or not.

aaronlehmann replied:
Do we also need to check if `s.Endpoint` is nil? There is a check for that below, but if it's really possible for `s.Endpoint` to be nil, this would panic.

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -186,12 +187,28 @@ func (na *NetworkAllocator) ServiceAllocate(s *api.Service) (err error) {
+			if nAttach.Target == eAttach.NetworkID {
+				match = true
+				if err = na.allocateVIP(eAttach); err != nil {
+					return
```
Unrelated to the change, but let's make this `return err`. Otherwise, if someone uses `err := ...` in one of the scopes this belongs to, the `err` variable being set isn't the return value, and the function would accidentally return nil.

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -186,12 +187,28 @@ func (na *NetworkAllocator) ServiceAllocate(s *api.Service) (err error) {
+	var match bool
+	for idx, eAttach := range s.Endpoint.VirtualIPs {
+		match = false
+		for _, nAttach := range s.Spec.Task.Networks {
```
Instead of accessing `s.Spec.Task.Networks` directly, move the following code from below above this, and use `specNetworks`:

```go
        // Always prefer NetworkAttachmentConfig in the TaskSpec
        specNetworks := s.Spec.Task.Networks
        if len(specNetworks) == 0 && s != nil && len(s.Spec.Networks) != 0 {
                specNetworks = s.Spec.Networks
        }
```

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -186,12 +187,28 @@ func (na *NetworkAllocator) ServiceAllocate(s *api.Service) (err error) {
+				if err = na.allocateVIP(eAttach); err != nil {
+					return
+				}
+				eVIPs = append(eVIPs, s.Endpoint.VirtualIPs[idx])
```
Can this be:

```go
eVIPs = append(eVIPs, eAttach)
```

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -186,12 +187,28 @@ func (na *NetworkAllocator) ServiceAllocate(s *api.Service) (err error) {
-	for _, eAttach := range s.Endpoint.VirtualIPs {
-		if err = na.allocateVIP(eAttach); err != nil {
-			return
+	// Allocate VIPs for all the pre-populated endpoint attachments
```
Also unrelated to the changes here, but I think it would be a great cleanup to split the contents of this function into `serviceAllocateVIP` and `serviceAllocateDNSRR` functions.

abhinandanpb commented:
@aaronlehmann addressed most of the comments minus couple of them
1) Invert function
2) Splitting DNSRR  and VIP allocation functions - I can address this in subsequent PR.

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -303,38 +324,81 @@ func OnInit(options *ServiceAllocationOpts) {
 func (na *NetworkAllocator) IsServiceAllocated(s *api.Service, flags ...func(*ServiceAllocationOpts)) bool {
-	var options ServiceAllocationOpts
+	return !na.ServiceNeedsAllocation(s, flags...)
+}
```
Please remove this function and change its callers to use `!na.ServiceNeedsAllocation(s, ...)` directly.

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -151,9 +151,10 @@ func (na *NetworkAllocator) Deallocate(n *api.Network) error {
 // IP and ports needed by the service.
-func (na *NetworkAllocator) ServiceAllocate(s *api.Service) (err error) {
+func (na *NetworkAllocator) ServiceAllocate(s *api.Service) error {
+	var err error
```
Go back to defining `err` in the function's return value. Otherwise, the `defer` only works properly when code sets this `err`, not some other variable named `err` in an inner scope. It's easy to accidentally break things.

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -199,6 +193,29 @@ func (na *NetworkAllocator) ServiceAllocate(s *api.Service) (err error) {
 
+	// Allocate VIPs for all the pre-populated endpoint attachments
+	eVIPs := s.Endpoint.VirtualIPs[:0]
+	var match bool
```
The loop below this does something similar, but uses a label and `continue` instead of a flag. I think the two loops should be structured consistently, but I don't have a strong preference about which which approach to use. I find the label/continue approach a little simpler, but I know there are people that don't like it.

If we go with the flag approach, this variable can be defined inside the for loop.

aaronlehmann commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -303,38 +324,81 @@ func OnInit(options *ServiceAllocationOpts) {
+
+		for _, net := range specNetworks {
+			match := false
+			for _, vip := range s.Endpoint.VirtualIPs {
```
`s.Endpoint` may be nil.

abhinandanpb replied:
I have added a check above which will return

aaronlehmann commented:
You will need to include a `Signed-off-by` line in the commits. See https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work and @GordonTheTurtle's comment at the top for details.

When you add this to the commit messages, please update the commit messages to include a paragraph or so that explains what you changed and why.

abhinandanpb commented:
@aaronlehmann : done --all

vieux commented:
ping @aluzzardi @aaronlehmann this one is needed for rc2, please take another look. thanks

aaronlehmann commented:
LGTM

I'll submit a followup for some of the cleanups I suggested.

dongluochen commented on manager/allocator/networkallocator/networkallocator.go:
```diff
@@ -300,41 +319,79 @@ func OnInit(options *ServiceAllocationOpts) {
+	//If the spec no longer has networks attached and has a vip allocated
+	//from previous spec the service needs to updated
+	if s.Endpoint != nil {
+		for _, vip := range s.Endpoint.VirtualIPs {
```
This nested loop is the same as line 349. You may want to keep them consistent (using the same code).

dongluochen replied:
If I'm not mistaken, they are doing the exact same check. I think it unnecessary to do it twice. You could cache the result, or reorganize the code to not duplicate.

abhinandanpb replied:
@dongluochen they are not  same. 349 scenario is when there are pre existing networks/newly added networks. This is hit when the spec has no networks but vips were allocated due to previous spec. The loop also checks for different things VIPs->Network vs Networks->Vips

dongluochen replied:
The`if` condition from outside is different. But from code perspective, `for X { for Y { check} }` makes the same check on the same data as `for Y { for X {check} }`. Using them with different order is actually confusing.

dongluochen replied:
Unless using this way would result in early termination of the loop.
